finality-provider/service/fp_instance.go

```diff
@@ -453,29 +319,32 @@ func (fp *FinalityProviderInstance) reportCriticalErr(err error) {
 	}
 }
 
-// checkLagging returns true if the lasted voted height is behind by a configured gap
-func (fp *FinalityProviderInstance) checkLagging(currentBlock *types.BlockInfo) bool {
-	return currentBlock.Height >= fp.GetLastProcessedHeight()+fp.cfg.FastSyncGap
-}
-
-// retrySubmitFinalitySignatureUntilBlockFinalized periodically tries to submit finality signature until success or the block is finalized
+// retrySubmitFinalitySignaturesUntilBlocksFinalized periodically tries to submit finality signature until success or the block is finalized
 // error will be returned if maximum retries have been reached or the query to the consumer chain fails
-func (fp *FinalityProviderInstance) retrySubmitFinalitySignatureUntilBlockFinalized(targetBlock *types.BlockInfo) (*types.TxResponse, error) {
+func (fp *FinalityProviderInstance) retrySubmitFinalitySignaturesUntilBlocksFinalized(targetBlocks []*types.BlockInfo) (*types.TxResponse, error) {
 	var failedCycles uint32
+	targetHeight := targetBlocks[len(targetBlocks)-1].Height
 
 	// we break the for loop if the block is finalized or the signature is successfully submitted
 	// error will be returned if maximum retries have been reached or the query to the consumer chain fails
 	for {
 		select {
 		case <-time.After(fp.cfg.SubmissionRetryInterval):
 			// error will be returned if max retries have been reached
-			res, err := fp.SubmitFinalitySignature(targetBlock)
+			var res *types.TxResponse
+			var err error
+			if len(targetBlocks) == 1 {
+				res, err = fp.SubmitFinalitySignature(targetBlocks[0])
```

Comment:
we can probably get rid of `SubmitFinalitySignature` from the interface

Reply:
I would keep the interface but the implementation could just use the one from the batch submission